core-bundle/src/Routing/Candidates/AbstractCandidates.php

```diff
@@ -0,0 +1,161 @@
+
+class AbstractCandidates implements CandidatesInterface
+{
+    /**
+     * A limit to apply to the number of candidates generated.
+     *
+     * This is to prevent abusive requests with a lot of "/". The limit is per
+     * batch, that is if a locale matches you could get as many as 2 * $limit
+     * candidates if the URL has that many slashes.
+     *
+     * @var int
+     */
+    private const LIMIT = 20;
+
+    /**
+     * @var array
+     */
+    private $urlPrefixes;
+
+    /**
+     * @var array
+     */
+    private $urlSuffixes;
+
+    public function __construct(array $urlPrefixes, array $urlSuffixes)
+    {
+        $this->urlPrefixes = $urlPrefixes;
+        $this->urlSuffixes = $urlSuffixes;
+    }
+
+    public function isCandidate($name): bool
+    {
+        return 0 === strncmp($name, 'tl_page.', 8);
+    }
+
+    public function restrictQuery($queryBuilder): void
+    {
+    }
+
+    /**
+     * Generates possible page aliases from the request path by
+     * removing prefixes, suffixes and parameters.
+     *
+     * Example 1:
+     *   Path: /en/alias/foo/bar.html
+     *   Prefixes: [en, de]
+     *   Suffixes: [.html]
+     *   Possible aliases:
+     *     - alias/foo/bar
+     *     - alias/foo
+     *     - alias
+     *
+     * Example 2:
+     *   Path: /en/alias/foo/bar.html
+     *   Prefixes: [en, '']
+     *   Suffixes: [.html, '']
+     *   Possible aliases:
+     *     - en/alias/foo/bar.html
+     *     - en/alias/foo/bar
+     *     - en/alias/foo
+     *     - en/alias
+     *     - alias/foo/bar.html
+     *     - alias/foo/bar
+     *     - alias/foo
+     *     - alias
+     */
+    public function getCandidates(Request $request): array
+    {
+        $url = $request->getPathInfo();
+        $url = rawurldecode(ltrim($url, '/'));
+        $candidates = [];
+
+        if (empty($url)) {
+            throw new \RuntimeException(__METHOD__.' cannot handle empty path');
+        }
+
+        foreach ($this->urlPrefixes as $prefix) {
+            // Language prefix only (e.g. URL = /en/)
+            if ($url === $prefix.'/') {
+                $candidates[] = 'index';
+                continue;
+            }
+
+            $withoutPrefix = $url;
+
+            if ('' !== $prefix) {
+                if (0 !== strncmp($url, $prefix.'/', \strlen($prefix) + 1)) {
+                    continue;
+                }
+
+                $withoutPrefix = substr($url, \strlen($prefix) + 1);
+            }
+
+            foreach ($this->urlSuffixes as $suffix) {
+                $withoutSuffix = $withoutPrefix;
+
+                if ('' !== $suffix) {
+                    if (0 !== substr_compare($withoutPrefix, $suffix, -\strlen($suffix))) {
+                        continue;
+                    }
+
+                    $withoutSuffix = substr($withoutPrefix, 0, -\strlen($suffix));
+                }
+
+                $this->addCandidatesFor($withoutSuffix, $candidates);
+            }
+        }
```

Comment:
What about some regex magic? :smile:  (I did only a quick test, not sure 

```suggestion
    foreach ($urlPrefixes as $prefix) {
        // Language prefix only (e.g. URL = /en/)
        if ($url === $prefix . '/') {
            $candidates[] = 'index';
            continue;
        }

        foreach ($urlSuffixes as $suffix) {
            // split into <optional prefix><middle part><optional suffix>
            $pattern = sprintf('#^(?:%s/?)?(\S*?)(?:%s)?$#', preg_quote($prefix, '#'), preg_quote($suffix, '#'));
            preg_match($pattern, $url, $matches);
            [, $withoutPrefixAndSuffix] = $matches;

            addCandidatesFor($withoutPrefixAndSuffix, $candidates);
        }
    }
```

Reply:
@m-vo Maybe we can optimize this in a second step and a separate PR? This one is already pretty complex.